sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,225 @@
+
+    async def _dispatch_notification_unlimited(self, n, device, context):
+        p256dh = device.pushkey
+        if not isinstance(device.data, dict):
+            logger.info("device.data is not a dict, reject pushkey")
+            return [device.pushkey]
+
+        endpoint = device.data.get("endpoint")
+        auth = device.data.get("auth")
+
+        if not p256dh or not endpoint or not auth:
+            logger.info("subscription info missing, reject pushkey")
+            return [device.pushkey]
+
+        subscription_info = {
+            "endpoint": endpoint,
+            "keys": {"p256dh": p256dh, "auth": auth},
+        }
+        payload = WebpushPushkin._build_payload(n, device)
+        data = json.dumps(payload)
+
+        # we use the semaphore to actually limit the number of concurrent
+        # requests, since the HTTPConnectionPool will actually just lead to more
+        # requests being created but not pooled – it does not perform limiting.
+        with QUEUE_TIME_HISTOGRAM.time():
+            with PENDING_REQUESTS_GAUGE.track_inprogress():
+                await self.connection_semaphore.acquire()
+
+        try:
+            with SEND_TIME_HISTOGRAM.time():
+                with ACTIVE_REQUESTS_GAUGE.track_inprogress():
+                    logger.info("sending payload %s", data)
+                    response_wrapper = webpush(
+                        subscription_info=subscription_info,
+                        data=data,
+                        vapid_private_key=self.vapid_private_key,
+                        vapid_claims=self.vapid_claims,
+                        requests_session=self.http_agent_wrapper,
+                    )
+                    response = await response_wrapper.deferred
+                    await readBody(response)
+        finally:
+            self.connection_semaphore.release()
+
+        # assume 4xx is permanent and 5xx is temporary
+        if 400 <= response.code < 500:
+            return [device.pushkey]
+        return []
+
+    @staticmethod
+    def _build_payload(n, device):
+        """
+        Build the payload data to be sent.
+        Args:
+            n: Notification to build the payload for.
+            device (Device): Device information to which the constructed payload
+            will be sent.
+
+        Returns:
+            JSON-compatible dict
+        """
+        payload = {}
+
+        default_payload = device.data.get("default_payload")
+        if isinstance(default_payload, dict):
+            payload.update(default_payload)
+
+        for attr in [
+            "room_id",
+            "room_name",
+            "room_alias",
+            "membership",
+            "event_id",
+            "sender",
+            "sender_display_name",
+            "user_is_target",
+            "type",
+            "content",
+        ]:
+            if getattr(n, attr, None):
+                payload[attr] = getattr(n, attr)
```

Comment:
Instead of calling getattr a few times, this is a bit simpler.

```suggestion
            value = getattr(n, attr, None)
            if value:
                payload[attr] = value
```